README.md
<p align='center'>
  <im src='/images/logo.png' width='200'>
<p>
  
# pw-gen

[![Downloads](https://pepy.tech/badge/pw-gen)](https://pepy.tech/project/pw-gen) ![Version](https://img.shields.io/badge/version-0.1.1-blue) ![Python-Version](https://img.shields.io/badge/python-3.9-blue) ![issues](https://img.shields.io/github/issues/lunAr-creator/pw-gen) ![PyPI - Status](https://img.shields.io/pypi/status/pw-gen) ![License](https://img.shields.io/github/license/lunAr-creator/pw-gen) 


```pw-gen``` is a library for dealing with generating secure randomised passwords that are customisable and strong simultaneously.
<br></br>
## Installation

Use the package manager [pip](https://pypi.org/project/pip/) to install pw-gen.

```bash
pip install pw-gen
```
#### If you are still using the old version of ```pw-gen``` and want to update it:

```bash
pip install --upgrade pw-gen
```
<br></br>
> **_NOTE:_**  If you are still using older versions of **pw-gen** please download the latest version as examples in this README will likely be incorrect for old versions. Thankyou.

<br></br>
## Basic usage

### Simple passwords (Less Arguments)
```python
from pw_gen import Simple

var = Simple(20)

print(var.generate())
print(var.result())
```

#### Output (Please note that output varies depending on arguments provided)

```
pcWW1QjppIWkzErqjdh8
```
<br></br>
### Complex passwords (More Arguments)

```python
from pw_gen import Complex

var = Complex(20, 'both', include_numbers=True, include_special_chars=True)

print(var.generate())
print(var.result())
```

#### Output (Please note that output varies depending on arguments provided)

```
\{=~#YR>XR@N+Q3K{WFB
```

<br></br>

### Memorable passwords (Easy to remember)

```python
from pw_gen import Memorable

var = Memorable(True)

print(var.generate())
print(var.result())
```

#### Output (Please note that output varies depending on arguments provided)

```
CobrandFlint0825
```

<br></br>
## Creating passwords - An in depth explantation

<details>
<summary>Creating a password</summary>
  
<br>

To customise and generate our password we must first create an instance of our password. This can be acheived by doing **var_name = type_of_password(args)**. This template can be used for all password types. At the moment, there are three varations of a password **Simple, Complex and Memorable**. 
  
 
Simple password require less arguements than a complex password, and it is also a base class that all other variations are derived from. To make a **Simple** password, we can assign to parameters: one of which is mandatory and the other one is optional. The first parameter is password length. This should be an integer. The second one is characters. It defaults to a string of ascii_letters and ascii_digits. However, you can overwrite this by specifying your own as a **string**. Example of how to create a **Simple** password:

```python
from pw_gen import Simple

var1 = Simple(20) # Specifying password length to 20 and characters will default to letters and numbers
var2 = Simple(20, 'abcdefghijklmnopqrstuvwxyz') # Specifying password length to 20 and characters will be set to the ones specified.
```

Complex passwords require 2 mandatory parameters and 2 optional parameters. Param 1 is password length (an int), param 2 is string_method. **string_method** refers to **upper** (upper case), **lower** (lowercase) and **both** (uppercase and lowercase). These arguements should be **strings**. The last two parameters are **include_numbers** (defaults to **True**) and **include_special_chars** (defaults to **False**). These are **keyword-only** parameters. They can be set to **True** or **False**. Therefore, they must be explicitly stated. E.g ```arg=bool``` Example of how to create a **Complex** password:

```python
from pw_gen import Complex

var = Complex(20, 'both', include_numbers=True, include_special_chars=True)
```
Lastly, a Memorable password is a password that can be easily remembered. It uses 2 random words from the ```random_word``` library. It then gets 3-4 random numbers and adds them to the end of the password. We can create a **Memorable** password by assigning 1 parameter which is **include_numbers** (this defaults to **True**). Example of how to create a **Memorable** password:

```python
from pw_gen import Memorable

var = Memorable()
```

</details>

<details>
<summary>Generating a password(s)</summary>
  
<br>

To generate a password we have to use the 'generate' method with our object. The generate method requires no parameters. To generate a password, we will first have to create an object (see **Creating a password**). We then use the ```generate``` method as seen in the example below:

```python
from pw_gen import {insert password type}

var = {insert password type}(args)
var.generate()

# or

print(var.generate)
```

To generate **muliple** passwords we can use a for loop:

```python
from pw_gen import {insert password type}

var = {insert password type}(args)
for i in range(INSERT NUM):
  print(var.generate())
```
<br></br>
> _**Note**_: If you generate one password (_password1_), and then generate another password (_password2_) with the same object, **_password1_** will be cleared and replaced by **_password2_**. Therefore,  if you would like to **keep** a password, simply append it to a list of passwords or or store them in a file, database etc.

</details>

<details>
<summary>Returning a password</summary>

<br>

To return a generated password we can simple use the ```result()``` method:

```python
from pw_gen import {insert password type}

var = {insert password type}(args)
print(var.generate())
print(var.result())
```

</details>

<br></br>

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License
https://opensource.org/licenses/MIT

## PYPI page:
https://pypi.org/project/pw-gen/


CHANGELOG.txt
Change Log
==========

0.0.1 (13/03/2021)
-------------------
- First Release

0.0.2 (24/03/2021)
-------------------
- Added additional functionality

0.0.3 (25/03/2021)
-------------------
- Third Release

0.0.4 (25/03/2021)
-------------------
- Fourth Release

0.0.5 (25/03/2021)
-------------------
- Added Pin class

0.0.6 (25/03/2021)
-------------------
- Refactored code
- Removed Pin class
- Added Random-Word library

0.0.7 (25/03/2021)
-------------------
- Included PyYAML as a requirement (forgot to add it)

0.0.8 (25/03/2021)
-------------------
- Fixed bug for Memorable passwords

0.0.9 (3/04/2021)
-------------------
- Added shuffling for simple and complex passwords in the generate method

0.1.0 (22/05/2021)
-------------------
- Removed shuffling (string_utils) and random_word library.
- Changed to using urllib.request instead of random_word

0.1.1 (22/05/2021)
-------------------
- Updated install_requires in setup.py


test.py
from collections import Counter
from matplotlib import pyplot as plt
import pw_gen

def main():
    """
        main function

        Args: none

        Example Usage: none
    """

    pw_list = []

    pw_test_lengths = [5, 6]

    pin_list = []

    pin_test_lengths = list(range(19, 20))

    pw_separator = ''

    sample_count = 1000000

    print("Password test lengths: ", pw_test_lengths)

    print("Pin Test Lengths: ", pin_test_lengths)

    for pin_test in pin_test_lengths:
        pin_test_filename = "pin_test_" + str(pin_test) + ".txt"

        print("Test filename for pins of length ", str(pin_test), ": ", pin_test_filename)

    for pw_test in pw_test_lengths:
        pw_test_filename = "pw_test_simple_" + str(pw_test) + ".txt"

        print("Test filename for simple passwords of length ", str(pw_test), ": ", pw_test_filename)

        for _ in range(sample_count):
            new_password = pw_gen.Simple(pin_test)
            pw_generated_list = new_password.generate()
            pw_final = new_password.result()

            if _ == 1:
                print("Simple Password Generation: ", new_password)
                print("Simple Password yielded: ", pw_final)

            del new_password

            pw_list.append(pw_final)

        with open(pw_test_filename, 'w') as file_handle:
            file_handle.write(pw_separator.join(pw_list))

        with open(pw_test_filename, 'r') as file_handle:
            sample_string = file_handle.read()

        print("1,000,000 passwords at ", str(pw_test), \
                " characters each yields: ", len(sample_string))

        counter_object = Counter(sample_string)

        counter_list = counter_object.most_common()

        print(counter_list)

        image_filename = pw_final + ".png"

        plt.scatter(*zip(*counter_list))
        plt.savefig(image_filename)
        plt.show()

    return 0

if __name__ == "__main__":
    main()


__init__test.py
import secrets
import string

class simplepass:
    def __init__(self, length: int, characters: str) -> None:
        self.length = length
        self.characters = characters

    def generate(self, iterations: int):
        '''
        Generates a password depending on length, given characters and iterations.
        Uses 2 for loops to achieve this
        '''
        for p in range(iterations):
            output_password = ''
            for c in range(self.length):
                output_password += secrets.choice(self.characters)
            print(output_password)

class complexpass(simplepass):
    def __init__(self, length, string_method, numbers=True, special_chars=False):
        self.length = length
        self.string_method = string_method
        self.numbers = numbers
        self.special_chars = special_chars

    def generate(self, iterations):
        '''
        Generates a string for the var characters that is used to make
        a password depending on length, upper of lowercase, digits, special
        characters and iterations. Uses 2 for loops to achieve this.
        '''

        characters = ''

        methods = {
            'upper': string.ascii_uppercase,
            'lower': string.ascii_lowercase,
            'both': string.ascii_letters
        }

        characters += methods[self.string_method]

        if self.numbers:
            characters += string.digits
        if self.special_chars:
            characters += string.punctuation

        # Generates the password

        for p in range(iterations):
            output_password = ''
            for c in range(self.length):
                output_password += secrets.choice(characters)
            print(output_password)

# Tests

if __name__ == '__main__':
    # Generate a complex password

    var1 = complexpass(20, 'both', True, True)
    var1.generate(3)

    # Generate a simple password

    var2 = simplepass(20, 'abcdefghijklmnopqrstuvwxyz0123467589')
    var2.generate(3)


pass_gen/init.py
import hashlib, binascii, os
import secrets
import string

password_result = []

class simple:
    def __init__(self, length: int, characters: str) -> None:
        self.length = length
        self.characters = characters

    def generate(self, iterations):
        password_result.clear()
        for p in range(iterations):
            password: str = ''
            for c in range(self.length):
                password += secrets.choice(self.characters)
            password_result.append(password)
        return password_result

    def result(self, num: int) -> None:
        return password_result[num]

class complex(simple):
    def __init__(self, length, string_method, numbers=True, special_chars=False):
        characters = ''

        methods: dict = {
            "upper": string.ascii_uppercase,
            "lower": string.ascii_lowercase,
            "both": string.ascii_letters,
        }


        characters += methods[string_method]

        if numbers:
            characters += string.digit
        if special_chars:
            characters += string.punctuation

        super().__init__(length=length, characters=characters)

def hash(password):
    salt = hashlib.sha256(os.urandom(60)).hexdigest().encode('ascii')
    pwdhash = hashlib.pbkdf2_hmac('sha512', password.encode('utf-8'),salt, 100000)

    pwdhash = binascii.hexlify(pwdhash)
    return (salt + pwdhash).decode('ascii')

def verify_hash(stored_password, provided_password):
    salt = stored_password[:64]
    stored_password = stored_password[64:]

    pwdhash = hashlib.pbkdf2_hmac('sha512',provided_password.encode('utf-8'),salt.encode('ascii'),100000)
    pwdhash = binascii.hexlify(pwdhash).decode('ascii')
    return pwdhash == stored_password

# Test scenarios

if __name__ == "__main__":
    var = simple(20, 'abcdefghijklmnopqrstuvwxyz0123467589')
    print(var.generate(3))

    print(var.result(1))

    stored_password = hash(var.result(1))
    print(stored_password)

    p = input('Password: ')
    if verify_hash(stored_password, p) == True:
        print('The password is correct')


pw_gen/__init__.py
import random
import secrets
import string

import urllib.request

class Simple():
    def __init__(self, length: int, *, characters = None):
        '''A simple password (less arguments compared to complex)'''
        self.length = length
        self.characters = characters
        self.output = ''

    def generate(self):
        '''
        Generates a password depending on the num_of_passwords
        and the arguments provided in the simple class
        '''
        characters = ''

        if self.characters is None:
            characters = string.ascii_letters + string.digits
        else:
            characters = self.characters

        self.output = ''
        password = ''

        for i in range(self.length):
            password += secrets.choice(characters)
        self.output += password

    def result(self):
        return str(self.output.__str__())

class Complex(Simple):
    def __init__(self, length, string_method, *, include_numbers=True, include_special_chars=False):
        '''
        Creates a customisable password depending on length, string_method, numbers and special_chars
        '''
        characters = ''
        self.output = ''

        methods: dict = {
            "upper": string.ascii_uppercase,
            "lower": string.ascii_lowercase,
            "both": string.ascii_letters,
        }

        characters += methods[string_method]

        if include_numbers:
            characters += string.digits
        if include_special_chars:
            characters += string.punctuation

        super().__init__(length=length, characters=characters)

class Memorable(Simple):
    def __init__(self, include_numbers=True):
        '''A memorable password e.g HelloWorld123'''
        self.include_numbers = include_numbers
        self.output = ''

    def generate(self):
        '''
        Generates the password containing 2 words and numbers if self.numbers == True
        '''
        word_url = "http://svnweb.freebsd.org/csrg/share/dict/words?view=co&content-type=text/plain"
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64)'}
        req = urllib.request.Request(word_url, headers=headers)
        response = response = urllib.request.urlopen(req)
        long_txt = response.read().decode()
        words = long_txt.splitlines()

        self.output = ''
        password = ''
        two_words = ''

        for i in range(2):
            two_words += secrets.choice(words).title()
            password = two_words

        if self.include_numbers:
            for i in range(random.randint(3, 4)):
                password += secrets.choice(string.digits)
            self.output += password

        return self.output

    def result(self):
        return str(self.output.__str__())


pw_gen_demo_and_examples.py
from pw_gen import Simple, Complex, Memorable, Pin

# ---------------------------------------------------------------------------- #
# ----------------------------Simple-Password-Tests--------------------------- #
# ---------------------------------------------------------------------------- #

password1 = Simple(20, 'abcdefghijklmnopqrstuvwxyz0123456789')
print(password1.generate(3))
print(password1.return_result(1))
password1.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['j81j14nujfaeit3fppkd', 'tn1m6qbg7vrt11a71qsr', 'k54s1ix3x7vku4vympno']
# tn1m6qbg7vrt11a71qsr
# ---------------------------------OUTPUT------------------------------------- #

password1 = Simple(20, 'abcdefghijklmnopqrstuvwxyz0123456789')
print(password1.generate(3))
print(password1.return_result(1))
password1.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['sfmraiaimdrl743o7z5j', '2e2m2qr6e4jd03hev4fd', 'j6l45ojcjfjude71ghy9']
# 2e2m2qr6e4jd03hev4fd
# ---------------------------------OUTPUT------------------------------------- #




# ---------------------------------------------------------------------------- #
# ---------------------------Complex-Password-Tests--------------------------- #
# ---------------------------------------------------------------------------- #

password2 = Complex(20, 'lower', False, True)
print(password2.generate(3))
print(password2.return_result(1))
password2.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['/$i=pigz/b[b`rr`+\\_g', ']z^cwe.o;@jvo;k+"zaa', "ovi@?]ttt'_\\f:j=+#*p"]
# ]z^cwe.o;@jvo;k+"zaa
# ---------------------------------OUTPUT------------------------------------- #

password2 = Complex(20, 'upper', True, False)
print(password2.generate(3))
print(password2.return_result(1))
password2.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['LSK0OU29NJCA93FPPK9D', 'SYCEEJJI1WURUZC42I43', 'VHJY8CTELMMOUZNSEB0N']
# SYCEEJJI1WURUZC42I43
# ---------------------------------OUTPUT------------------------------------- #

password2 = Complex(20, 'both', True, True)
print(password2.generate(3))
print(password2.return_result(1))
password2.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['G+dLU[+N;K&:r.CEf?k2', 'Bi27W4L41Zq6S[vlVya\\', '9("KF5#;H>[W"p77>uI{']
# Bi27W4L41Zq6S[vlVya\
# ---------------------------------OUTPUT------------------------------------- #



# ---------------------------------------------------------------------------- #
# -------------------------Memorable-Password-Tests--------------------------- #
# ---------------------------------------------------------------------------- #

password3 = Memorable()
print(password3.generate(3))
print(password3.return_result(1))
password3.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['PelicanDaffy5333', 'SidewalkRetrofitting685', 'TramwaySuppressor816']
# SidewalkRetrofitting685
# ---------------------------------OUTPUT------------------------------------- #

password3 = Memorable(True)
print(password3.generate(3))
print(password3.return_result(1))
password3.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['ShayHarvey097', 'ToggingConfect1648', 'HaloBernini219']
# ToggingConfect1648
# ---------------------------------OUTPUT------------------------------------- #

password3 = Memorable(False)
print(password3.generate(3))
print(password3.return_result(1))
password3.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['AcculturateProtagonist', 'UnderclassmenFiddle', 'EverymanIntimacy']
# UnderclassmenFiddle
# ---------------------------------OUTPUT------------------------------------- #




# ---------------------------------------------------------------------------- #
# ---------------------------Pin-Password-Tests------------------------------- #
# ---------------------------------------------------------------------------- #

password4 = Pin(4)
print(password4.generate(3))
print(password4.return_result(1))
password4.clear_results()

# ---------------------------------OUTPUT------------------------------------- #
# ['4854', '5587', '1374']
# 5587
# ---------------------------------OUTPUT------------------------------------- #


password_generator.py
import hashlib, binascii, os
import secrets
import string

password_result = []

class simple:
    def __init__(self, length: int, characters: str) -> None:
        self.length = length
        self.characters = characters

    def generate(self, iterations):
        password_result.clear()
        for p in range(iterations):
            password: str = ''
            for c in range(self.length):
                password += secrets.choice(self.characters)
            password_result.append(password)
        return password_result

    def result(self, num: int) -> None:
        return password_result[num]

class complex(simple):
    def __init__(self, length, string_method, numbers=True, special_chars=False):
        characters = ''

        methods: dict = {
            "upper": string.ascii_uppercase,
            "lower": string.ascii_lowercase,
            "both": string.ascii_letters,
        }
       

        characters += methods[string_method]

        if numbers:
            characters += string.digit
        if special_chars:
            characters += string.punctuation

        super().__init__(length=length, characters=characters)

def hash(password):
    salt = hashlib.sha256(os.urandom(60)).hexdigest().encode('ascii')
    pwdhash = hashlib.pbkdf2_hmac('sha512', password.encode('utf-8'),salt, 100000)

    pwdhash = binascii.hexlify(pwdhash)
    return (salt + pwdhash).decode('ascii')

def verify_hash(stored_password, provided_password):
    salt = stored_password[:64]
    stored_password = stored_password[64:]

    pwdhash = hashlib.pbkdf2_hmac('sha512',provided_password.encode('utf-8'),salt.encode('ascii'),100000)
    pwdhash = binascii.hexlify(pwdhash).decode('ascii')
    return pwdhash == stored_password

# Test scenarios

if __name__ == "__main__":
    var = simple(20, 'abcdefghijklmnopqrstuvwxyz0123467589')
    print(var.generate(3))

    print(var.result(1))

    stored_password = hash(var.result(1))
    print(stored_password)

    p = input('Password: ')
    if verify_hash(stored_password, p) == True:
        print('The password is correct')

